Description: Update the includes for the new glib version
Forwarded: not-needed
Author: Alexis Bezverkhyy <alexis-externe.bezverkhyy@edf.fr>
Last-Update: 2015-08-18

Index: strongswan-4.5.2/src/libcharon/plugins/nm/nm_service.h
===================================================================
--- strongswan-4.5.2.orig/src/libcharon/plugins/nm/nm_service.h	2010-10-22 16:33:30.000000000 +0200
+++ strongswan-4.5.2/src/libcharon/plugins/nm/nm_service.h	2015-08-18 15:59:43.650304865 +0200
@@ -21,8 +21,7 @@
 #ifndef NM_SERVICE_H_
 #define NM_SERVICE_H_
 
-#include <glib/gtypes.h>
-#include <glib-object.h>
+#include <glib.h>
 #include <nm-vpn-plugin.h>
 
 #include "nm_creds.h"
